### PR TITLE
VOL-2180: Introducing AWS Cognito User Migration Lambda (LDAP)

### PR DESCRIPTION
**Pull request by jerotire**
## Description

Introduces the User Migration Lambda for use by AWS Cognito. The lambda accepts an event from AWS Cognito (forgotten password or authentication) and attempts to either fetch a user (forgotten password journey) or authenticates a user (authentication journey) against an LDAP service. The response from the lambda will instruct AWS Cognito to create a user (migrate them) or throw an exception (User Not Found or Invalid Credentials).

Related issue: VOL-2180

## Type of change
Feature (non-breaking change which adds functionality)

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by sdh100shaun — APPROVED** [review]

File: src/util/logger.ts [hunk 1]
```diff
@@ -7,6 +7,10 @@ export class Logger {
     this.logFormat = `{ "apiRequestId": "${apiRequestId}", "correlationId": "${correlationId}", "message": "%s" }`;
   }
 
+  public trace(msg: string, ...args): void {
```
Comment (Hodgy) [inline src/util/logger.ts:10]: Worth pulling this into the starter as well?

File: template.yaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+          LDAP_ADMIN_PASSWORD: ''
+          LDAP_USER_SEARCH_BASE: ''
+          LDAP_USERNAME_ATTRIBUTE: ''
+          LDAP_OBJECT_FILTER_MAP: ''
+
+Outputs:
+  CognitoUserMigrationFunction:
+    Description: "Cognito User Migration Lambda Function ARN"
+    Value: !GetAtt HelloWorldFunction.Arn
+  HelloWorldFunctionIamRole:
```
Comment (Hodgy) [inline template.yaml:24]: Name?

File: template.yaml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+          LDAP_ADMIN_DN: ''
+          LDAP_ADMIN_PASSWORD: ''
+          LDAP_USER_SEARCH_BASE: ''
+          LDAP_USERNAME_ATTRIBUTE: ''
+          LDAP_OBJECT_FILTER_MAP: ''
+
+Outputs:
+  CognitoUserMigrationFunction:
+    Description: "Cognito User Migration Lambda Function ARN"
+    Value: !GetAtt HelloWorldFunction.Arn
```
Comment (Hodgy) [inline template.yaml:23]: Name?

Comment (jerotire) [reply template.yaml:23]: Nice spot.

File: template.yaml (hunk 2 above)
Comment (jerotire) [reply template.yaml:24]: Nice spot 😄

File: event/userMigrationAuthenticationTriggerEvent.event.json [hunk 4]
```diff
@@ -0,0 +1,12 @@
+{
+  "triggerSource": "UserMigration_Authentication",
+  "userName": "test",
+  "request": {
+    "password": "test"
+  },
+  "response": {
+    "userAttributes": {},
+    "finalUserStatus": "Nope",
```
Comment (JoshuaLicense) [inline event/userMigrationAuthenticationTriggerEvent.event.json:9]: Would these be better as valid values? `UNKNOWN` & `SUPPRESS`?

File: package-lock.json [hunk 5]
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "dvsa-lambda-starter",
+  "name": "cognito-user-migration",
   "version": "1.0.0",
```
Comment (JoshuaLicense) [inline package-lock.json:3]: Quite minor in this case, but using npm v7 (making this `version: "2.0.0"`) will give better performance ([source](https://user-images.githubusercontent.com/459713/105940574-a1d5c880-6029-11eb-8461-9b57bff84c08.png)) while doing _most_ npm commands.

File: src/service/ldapUserStoreService.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+export class LdapUserStoreService {
+  private readonly logger: Logger;
+
+  constructor(logger: Logger) {
+    this.logger = logger;
+  }
+
+  public async getUser(userName: string): Promise<Entry | null> {
+    const client = new Client({
+      url: process.env.LDAP_URL,
```
Comment (JoshuaLicense) [inline src/service/ldapUserStoreService.ts:15]: Would timeouts being specified be a good idea here - `timeout` and `connectTimeout`? The default to `timeout` is currently infinity.

**Review by JoshuaLicense — CHANGES_REQUESTED** [review]

File: src/util/logger.ts (hunk 1 above)
Comment (jerotire) [reply src/util/logger.ts]: Yup, I'll make a new PR for that on the Lambda Starter.

Comment (jerotire) [reply src/util/logger.ts]: Done. PR created (https://github.com/dvsa/dvsa-lambda-starter/pull/47)

File: package-lock.json (hunk 5 above)
Comment (jerotire) [reply package-lock.json:3]: Are we talking about the **version** or **lockfileVersion**? The **version** is just duplicated from the version defined in `package.json`.

Look at this, I think we mean **lockFileVersion**? That's just an integer, and NPM did that automattically.
https://docs.npmjs.com/cli/v7/configuring-npm/package-lock-json

I've installed NPM v7, to get the new format for the lock file.

Comment (JoshuaLicense) [reply package-lock.json:3]: Yup `lockfileVersion` - clicked on the wrong line.

File: src/service/ldapUserStoreService.ts (hunk 6 above)
Comment (jerotire) [reply src/service/ldapUserStoreService.ts:15]: I think we can benefit with a "timeout" being specified, maybe even with it being definable from `process.env`.

File: event/userMigrationAuthenticationTriggerEvent.event.json (hunk 4 above)
Comment (jerotire) [reply event/userMigrationAuthenticationTriggerEvent.event.json]: Yup, I'll remove them as they are set by the lambda.

**Review by JoshuaLicense — APPROVED** [review]

**Review by sdh100shaun — APPROVED** [review]

**Comment by sdh100shaun** [comment]
@jerotire  can we update the AWS sdk please see https://github.com/dvsa/cognito-migration-lambda/pull/2 and double-check before merge

**Comment by jerotire** [comment]
> @jerotire can we update the AWS sdk please see #2 and double-check before merge

Yup, only required for dev dependancies, not for Production.

With the NPM update, I also ran npm update, and we now have `"aws-sdk": "^2.969.0",` which surpasses the one defined by Snyk in #2.

**Review by Hodgy — APPROVED** [review]

**Pull request merged**